Content.Server/Power/EntitySystems/ApcSystem.cs

```diff
@@ -0,0 +1,181 @@
+        {
+            TryComp<AccessReaderComponent>(uid, out var access);
+            if (args.Session.AttachedEntity == null)
+                return;
+
+            if (access == null || _accessReader.IsAllowed(access, args.Session.AttachedEntity.Value))
+            {
+                component.MainBreakerEnabled = !component.MainBreakerEnabled;
+                Comp<PowerNetworkBatteryComponent>(uid).CanDischarge = component.MainBreakerEnabled;
+
+                UpdateUIState(uid, component);
+                SoundSystem.Play(Filter.Pvs(uid), component.OnReceiveMessageSound.GetSound(), uid, AudioParams.Default.WithVolume(-2f));
+            }
+            else
+            {
+                _popupSystem.PopupCursor(Loc.GetString("apc-component-insufficient-access"),
+                    Filter.Entities(args.Session.AttachedEntity.Value));
+            }
+        }
+
+        public void UpdateApcState(EntityUid uid,
+            ApcComponent? apc=null,
+            BatteryComponent? battery=null)
+        {
+            if (!Resolve(uid, ref apc, ref battery))
+                return;
+
+            var newState = CalcChargeState(uid, apc, battery);
+            if (newState != apc.LastChargeState && apc.LastChargeStateTime + TimeSpan.FromSeconds(ApcComponent.VisualsChangeDelay) < _gameTiming.CurTime)
+            {
+                apc.LastChargeState = newState;
+                apc.LastChargeStateTime = _gameTiming.CurTime;
+
+                if (TryComp(uid, out AppearanceComponent? appearance))
+                {
+                    appearance.SetData(ApcVisuals.ChargeState, newState);
+                }
+
+                if (TryComp(uid, out SharedPointLightComponent? light))
+                {
+                    light.Color = newState switch
+                    {
+                        ApcChargeState.Lack => ApcComponent.LackColor,
+                        ApcChargeState.Charging => ApcComponent.ChargingColor,
+                        ApcChargeState.Full => ApcComponent.FullColor,
+                        _ => ApcComponent.LackColor
+                    };
+                    light.Dirty();
+                }
+            }
+
+            bool uiDirty = false;
+            var newCharge = battery.CurrentCharge;
+            if (apc.LastChargeChange + TimeSpan.FromSeconds(ApcComponent.VisualsChangeDelay) < _gameTiming.CurTime)
+            {
+                apc.LastChargeChange = _gameTiming.CurTime;
+                uiDirty = true;
+            }
+
+            var extPowerState = CalcExtPowerState(uid, apc, battery);
+            if (extPowerState != apc.LastExternalState
+                && apc.LastExternalStateTime + TimeSpan.FromSeconds(ApcComponent.VisualsChangeDelay) < _gameTiming.CurTime)
+            {
+                apc.LastExternalState = extPowerState;
+                apc.LastExternalStateTime = _gameTiming.CurTime;
+                uiDirty = true;
+            }
```

Comment:
Maybe just combine `LastExternalState` and `LastChargeChange` into a `LastUiUpdate`?

regardless of which of these set uiDiryt = true, the UI gets sent and `CalcExtPowerState()` gets called.

Reply:
opted to just remove lastchargechange